analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -37,110 +37,157 @@ namespace SonarAnalyzer.Rules.CSharp
 
-        private static readonly ISet<SyntaxKind> notAllowedDisposeModifiers = new HashSet<SyntaxKind>
+        private static readonly ISet<SyntaxKind> NotAllowedDisposeModifiers = new HashSet<SyntaxKind>
         {
             SyntaxKind.VirtualKeyword,
             SyntaxKind.AbstractKeyword
         };
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
-            {
-                var classDeclaration = (ClassDeclarationSyntax)c.Node;
-
-                var checker = new DisposableChecker(classDeclaration, c.SemanticModel);
-                var locations = checker.GetIssueLocations();
-                if (locations.Any())
                 {
-                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
-                        locations.ToAdditionalLocations(),
-                        locations.ToProperties()));
-                }
-            },
-            SyntaxKind.ClassDeclaration);
-        }
+                    if (c.ContainingSymbol.Kind != SymbolKind.NamedType)
+                    {
+                        return;
+                    }
+
+                    DisposableChecker checker;
+                    SyntaxToken declarationIdentifier;
+                    if (c.Node.IsKind(SyntaxKind.ClassDeclaration))
+                    {
+                        var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                        declarationIdentifier = classDeclaration.Identifier;
+                        checker = new DisposableChecker(classDeclaration.BaseList, declarationIdentifier, c.SemanticModel.GetDeclaredSymbol(classDeclaration), "class", c.SemanticModel);
+                    }
+                    else
+                    {
+                        var recordDeclaration = (RecordDeclarationSyntaxWrapper)c.Node;
+                        declarationIdentifier = recordDeclaration.Identifier;
+                        checker = new DisposableChecker(recordDeclaration.BaseList, declarationIdentifier, c.SemanticModel.GetDeclaredSymbol(recordDeclaration), "record", c.SemanticModel);
+                    }
+
+                    var locations = checker.GetIssueLocations();
```

Comment:
`locations` is enumerable and it's iterated 3 times. We should update GetIssueLocations to return a List or array.